---
title: "Part 1 - Loading and preprocessing microarray data"
author: "Andreas Schüttler"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Part 1- Loading and preprocessing microarray data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The package `toxprofileR` mainly builds on the packages `limma`, `kohonen` and `hydromad` to supply functions for creating dynamic toxicogenomic fingerprints. The package currently only works with data from Agilent based zebrafish microarrays designed by Oaklabs.
This vignette describes how the data is (should be) generated and how the package can be used to load and preprocess the data to retrieve logFC data.

# Experiment
## Experimental design

Here could follow some explanation of our experimental design.

## Description of example experiment and substances

### Diuron
### Diclofenac
### Naproxen

# Analysis workflow
## Data import

First, we create a folder in our working directory called "data" and a folder for each substance tested, where we place the raw data-output from the feature-extraction software. 

For each substance we need a tab-separated target file in `.csv` format containing the following columns:

* `FileName`: Names of the raw data file
* `sample_ID`: ID of experimental sample
* `substance`: Name of substance
* `concentration_level`: "Control", "C1", "C2", ..., "Cn" for increasing concentrations applied
* `concentration_umol_l`: Concentration of applied substance in µmol/L
* `time_hpe`: Time of measurement after exposure
* `type`: "treatment", "control" or "recovery"
* `comment`: additional comments
* `scan_ID`: scanID (1 if there was only one scan)

Next we can use the function `importData` to import the raw data into R and remove outlier arrays based on their interquartile range. If the iqr of an array deviates more than two standard deviations from the population mean (of iqrs) it is removed. 

Here we import data from exposure experiments with the three substances diuron, diclofenac and naproxen.

```{r import data, fig.show='hold'}
library("toxprofileR")

for(substance in c("diuron","diclofenac","naproxen")){

message(substance)
targetfile = paste0("./data/", substance, "/targetsfile_", substance, ".csv")
datadir = paste0("./data/", substance, "/ArrayData/")

assign(paste("data",substance,sep = "_"),
       value =  toxprofileR::importData(targetfile = targetfile,
                        datadir = datadir,
                        output = T,
                        removeOutliers = T))
}
```


## Normalisation

Next, we normalise the data between the arrays. Here we normalize all three datasets together, using cyclic-loess normalisation. However normalisation of single datasets is also possible.

```{r normalise, echo=TRUE, results='asis'}
dslist <- list(data_diuron, data_diclofenac, data_naproxen)
dslist_norm <- toxprofileR::normalizeBatch(dslist = dslist)
```

## Further data preprocessing

In the next step some further preprocessing is applied to the data. Measurements of duplicate probes are averaged (at the same time removing probes which are flagged of poor quality), a recent annotation of the array is added and batch correction is applied, if necessary. In our example, this was necessary for the diclofenac experiment, since the arrays were processed in two different labs.

```{r preprocess}
data_proc_diuron <- toxprofileR::preprocess(dslist_norm[[1]], batchcorrect = F)

batch_diclofenac <- as.factor(as.numeric(dslist_norm[[2]]$targets$comment=="Oaklabscan"))
data_proc_diclofenac <- toxprofileR::preprocess(dslist_norm[[2]],
                                                batchcorrect = T,
                                                batch = batch_diclofenac)

data_proc_naproxen <- toxprofileR::preprocess(dslist_norm[[3]], batchcorrect = F)
```

## Quality control

In the end of data preprocessing we check the quality of the data with some analytical plots. We check the expression of control probes (dark corners, bright corners, spike ins) and the overall data distribution with a multidimensional scaling plot.

### Diuron
```{r qcplot diuron}
toxprofileR::qc_plots(data_proc_diuron)
```

### Diclofenac
```{r qc plot diclofenac}
toxprofileR::qc_plots(data_proc_diclofenac)
```

### Naproxen
```{r qc plot naproxen}
toxprofileR::qc_plots(data_proc_naproxen)
```

## Calculate logFC

In our experiment gene expression was measured at several time points. Since the embryo develops during that time there is a lot of change in gene expression due to development, we are not primarily interested in here. Therefore we normalize the expression values to the controls of each time point, therefore retrieving logFCs for each timepoint.

```{r logFC}
for(substance in c("diuron","diclofenac","naproxen")){
    assign(paste("data",substance,"logFC",sep = "_"),
           value = toxprofileR::calc_logfc(get(paste0("data_proc_",substance))))

    save(list = paste("data",substance,"logFC",sep = "_"), file = paste0("./data/",substance,"/data_logfc_",substance,".Rd"))
}
```

##Session Info
```{r session info, echo=FALSE}
devtools::session_info()
```
